### PR TITLE
dynamical-friction

### PR DESCRIPTION
**Pull request by nianyic7**
Implemented dynamical friction for BH with different options controlled by BH_DynamicalFriction in the parameter file. Still in the testing/debugging phase and minor changes need to be made.

### CODE REVIEW

File: libgadget/blackhole.c [hunk 1]
```diff
@@ -515,12 +653,60 @@ blackhole_accretion_preprocess(int n, TreeWalk * tw)
...
+        {
+            BHP(n).DFAccel[j] = - 4. * M_PI * All.G * All.G * P[n].Mass * rho_frac * 
+            log_lam * (P[n].Vel[j] - BH_GET_PRIV(tw)->BH_SurroundingVel[PI][j]) / pow(bhvel, 3);
+            P[n].GravAccel[j]  += All.BH_DFBoost * BHP(n).DFAccel[j]; // Add a boost factor
+        }
+    }
+    else
+    {
+        for(j = 0; j < 3; j++) 
+        {
```
Comment (sbird) [inline libgadget/blackhole.c:702]: Better would be to just not allocate the array if DynamicalFriction is off and avoid this else clause. On use one would check for a NULL pointer in this array instead. An alternative is to change this to an unconditional memset(DFAccel, 0, 3) before the if clause.

File: libgadget/blackhole.c [hunk 2]
```diff
@@ -611,6 +797,26 @@ blackhole_accretion_ngbiter(TreeWalkQueryBHAccretion * I,
         } while(!__atomic_compare_exchange_n(&(BHP(other).SwallowID), &readid, newswallowid, 0, __ATOMIC_RELAXED, __ATOMIC_RELAXED));
     }
 
+    /*************************************************************************/
+
+    /* Collect surrounding DM/Star (optional gas) density/velocity for DF computation */
+    if(P[other].Type == 1 || P[other].Type == 4 || (P[other].Type == 0 && All.BH_DynamicalFriction == 0) ){
```
Comment (sbird) [inline libgadget/blackhole.c:803]: At this point I realised that BH_DynamicalFriction == 0 means Dynamical Friction is ON and BH_DynamicalFriction == 3 means it is OFF! That's odd.

File: libgadget/blackhole.c [hunk 3]
```diff
@@ -722,6 +935,40 @@ blackhole_feedback_ngbiter(TreeWalkQueryBHFeedback * I,
...
+            for(k = 0; k < 3; k++) 
+            {
+                bhvel += pow(I->Vel[k] - I->SurroundingVel[k], 2);
+                othervel += pow(P[other].Vel[k] - I->SurroundingVel[k], 2);
+            }
+            bhvel = sqrt(bhvel);
+            othervel = sqrt(othervel);
+
+        
+            if (othervel < bhvel)
```
Comment (sbird) [inline libgadget/blackhole.c:962]: Don't really need the sqrt here

File: libgadget/drift.c [hunk 4]
```diff
@@ -47,7 +48,7 @@ static void real_drift_particle(int i, inttime_t ti1, const double ddrift, const
     /* Jumping of BH */
     if(P[i].Type == 5) {
         int k;
-        if (BHP(i).JumpToMinPot) {
+        if (BHP(i).JumpToMinPot && All.BH_DynamicalFriction == 3) {
```
Comment (sbird) [inline libgadget/drift.c:51]: Cleaner to just not set JumpToMinPot in blackhole.c if repositioning is off.

File: libgadget/init.c [hunk 5]
```diff
@@ -96,6 +96,8 @@ set_init_params(ParameterSet * ps)
 
         All.BlackHoleOn = param_get_int(ps, "BlackHoleOn");
         All.WriteBlackHoleDetails = param_get_int(ps,"WriteBlackHoleDetails");
+        All.BH_DynamicalFriction = param_get_int(ps,"BH_DynamicalFriction");
+        All.BH_DFBoost = param_get_int(ps,"BH_DFBoost");
```
Comment (sbird) [inline libgadget/init.c:100]: Could you add a separate extra parameter that enables and disables repositioning? Something like BH_Reposition? That way we can easily disable repositioning without turning on dynamical friction.

File: libgadget/blackhole.c [hunk 6]
```diff
@@ -819,10 +1068,15 @@ blackhole_accretion_reduce(int place, TreeWalkResultBHAccretion * remote, enum T
...
             BHP(place).MinPotPos[k] = remote->BH_MinPotPos[k];
-            /* We set the velocity as well because
-             * we don't want drift to move us too far.*/
-            P[place].Vel[k] = remote->BH_MinPotVel[k];
+            /*********************************************************************/
+            if (All.BH_DynamicalFriction == 3){ /* reset vel only when reposition is on*/
+                /* We set the velocity as well because
+                 * we don't want drift to move us too far.*/
+                P[place].Vel[k] = remote->BH_MinPotVel[k];
+            }
```
Comment (sbird) [inline libgadget/blackhole.c:1076]: This is a good catch: I think I will fix this in the master branch

**Comment by nianyic7** [comment]
I might have opened my PR too early and there are a few more changes I want to make to DF. I didn't know that after opening a PR everyone watching MP-Gadget will keep receiving notifications when I push... In order not to bother other people with my commits I will close this pull request for now and make another one after more testings/changes. Sorry about that!

**Comment by sbird** [comment]
Hi Nianyi - please don't apologise, I prefer it if you open the PR early. That way everything is in the open and we know about conflicts early, plus we can pull any bugfixes directly into master. eg, there are now conflicts with https://github.com/MP-Gadget/MP-Gadget/pull/446 . If we didn't want to receive push emails we can turn them off. Anyway, I get messages every time someone pushes to the LSST repos. If you are not ready for review you can just put "WIP:" in the PR name and we will ignore it.

File: libgadget/blackhole.c (hunk 2 above)
Comment (rainwoodman) [reply libgadget/blackhole.c:803]: In any case, BH_DynamicalFrictionMethod or Mode / Type is more specific than DynamiccalFriction alone.

**Pull request closed without merging**